confmap/confmap.go

```diff
@@ -158,7 +159,7 @@ func decodeConfig(m *Conf, result any, errorUnused bool) error {
 			mapKeyStringToMapKeyTextUnmarshalerHookFunc(),
 			mapstructure.StringToTimeDurationHookFunc(),
 			mapstructure.TextUnmarshallerHookFunc(),
-			unmarshalerHookFunc(result),
+			unmarshalerHookFunc(result, topLevelUnmarshaling),
 			// after the main unmarshaler hook is called,
 			// we unmarshal the embedded structs if present to merge with the result:
 			unmarshalerEmbeddedStructsHookFunc(),
```

Comment:
If I understand correctly, there is no need to have `self` living at the `confmap.Conf` level, this is just used by `decodeConfig`. Could we have some sort of helper struct that holds `self` and has the unmarshaling hooks as methods? I think that would be cleaner (keeps the logic contained within `decodeConfig` instead of spilling into `confmap.Conf` itself) and possibly simpler to understand

Reply:
I don't think that works. This change addresses the successive calls on `confmap.Conf`, being called recursively by `Unmarshaler`. The `confmap.Conf` object has to be stateful to keep track of its top level object.

Reply:
I am at a loss here, sorry. If `confmap.Conf` has to be stateful, then how does https://github.com/open-telemetry/opentelemetry-collector/pull/9750#discussion_r1525365294 also hold? Wouldn't we have to reset the state each time we call `Unmarshal`?

Reply:
It resets the state before calling `Unmarshal` on the `Unmarshaler` interface, yes.